### DISCUSSION TITLE
Improve i18n Config Flexibility

### DISCUSSION THREAD

**Discussion by wvanrensselaer** [vercel/next.js#32556] [Ideas]
### Describe the feature you'd like to request

The locale is currently tightly coupled to the request path which is not ideal in some situations. For example consider that we want:

- http://example.com (`en-US`)
- http://example.ca (`en-CA`)
- http://example.ca/fr (`fr-CA`)

Example config ([from my test repo](https://github.com/wvanrensselaer/next-i18n-test)):

```js
module.exports = {
  i18n: {
    locales: ['en-US', 'en-CA', 'fr-CA', 'fr'],
    defaultLocale: 'en-US',
    domains: [
      {
        domain: 'example.com',
        defaultLocale: 'en-US',
        locales: ['en-US']
      },
      {
        domain: 'example.ca',
        defaultLocale: 'en-CA',
        locales: ['en-CA', 'fr-CA', 'fr']
      }
    ]
  }
};
```

Unless I'm missing something in the [docs](https://nextjs.org/docs/advanced-features/i18n-routing), I can't seem to get `fr-CA` from http://example.ca/fr. The locale always comes through as `fr` alone. I'd have to use http://example.ca/fr-CA which is not ideal as `-CA` is redundant with the `.ca` domain. `fr` alone may not be ideal if you also support France as a region which might be a slightly different translation set from Canadian French.

Side note - Related to https://github.com/vercel/next.js/issues/23829, this allows http://example.com/fr and http://example.ca/en-US which is not ideal.

### Describe the solution you'd like

A couple solutions come to mind, but I wonder if people have thoughts on other solutions:

Support path mappings:

```js
module.exports = {
  i18n: {
    defaultLocale: 'en-US',
    domains: [
      {
        domain: 'example.ca',
        defaultLocale: 'en-CA',
        paths: {
          '/fr': 'fr-CA',
        },
      },
    ],
  },
};
```

Or decoupling region and language in the config and rejoining as locale in `useRouter`:

```js
module.exports = {
  i18n: {
    defaultRegion: 'US',
    defaultLanguage: 'en',
    domains: [
      {
        domain: 'example.ca',
        region: 'CA', // Region for domain
        language: 'en', // Language for domain
        languages: ['fr'], // Custom language at /<language>
      },
    ],
  },
};
```

This one, like the current config is opinionated about path structure which may not be ideal. I'd also need to think more about this config when you want multiple regions on the same domain (if you don't own a region-specific domain).

Also happy to contribute if we agree on a solution.

### Describe alternatives you've considered

[One workaround I tried](https://github.com/wvanrensselaer/next-i18n-test/blob/main/pages/_middleware.tsx) was a middleware redirect to redirect `/fr` to `/fr-CA`. This works but I'd rather keep the `/fr` pathing and avoid the overhead of a redirect.

The [other workaround I tried](https://github.com/wvanrensselaer/next-i18n-test/blob/main/next.config.js#L21) was a rewrite rule of `/fr` to `/fr-CA`. This resulted in a 404 on the `/fr` route which seems like a bug or maybe I'm configuring this incorrectly.

The other option of course is to just use `/fr` with the provided `fr` locale and tack on the region separately, but it would be mighty dandy if the config supported this out of the box.

**Comment by nicholas-c** [comment]
This is also something I'd like to see as we currently have business cases where the following is required.

URL structures would ideally be `ca.example.com`, `ca.example.com/fr`, `uk.example.com`, `www.example.com` and `www.example.com/es` as stated in the opening discussion, this would be best as you're not redefining the "region" in your path.

However our locales would be setup as `en-CA`, `fr-CA`, `en-GB`, `en-US` and `es-US`

With the our current config, only `ca.example.com/fr-CA` would work which gives URL path redundancies.

```
        locales: ['es-US', 'en-GB', 'en-US', 'en-CA', 'fr-CA'],
        defaultLocale: 'en-US',
        domains: [
            {
                domain: 'example.com',
                defaultLocale: 'en-US',
                locales: ['es-US'],
            },
            {
                domain: 'uk.example.com',
                defaultLocale: 'en-GB',
            },
            {
                domain: 'ca.example.com',
                defaultLocale: 'en-CA',
                locales: ['fr-CA'],
            },
            {
                domain: 'fr.example.com',
                defaultLocale: 'fr-FR',
                locales: ['fr-FR'],
            },
            {
                domain: 'es.example.com',
                defaultLocale: 'es-ES',
                locales: ['es-ES'],
            },
        ],
```

Ideally, solutionising, we'd still define the locales as the full `lang-Region` style, however this could be split and provided to the developer as `{ locale, language, region }` - Something `new Intl.Locale` could help with fully.

Further to this, multiple countries/regions can support the same language (Spanish and French, looking at you), however the content fed from the CMS may look completely different per region (Locale). Having locales defined as just `fr`, `es`, `en` etc simply wouldn't work here.

**Comment by jacobrask** [comment]
Paths mapping would be great, we also have an International site at /intl with the `en-001` locale, but don’t want `en-001` in the URL.
